pkg/utils/packet.go

```diff
@@ -200,22 +200,34 @@ func AnnounceIPs(ifName string, ipConfigs []*current.IPConfig) error {
 	}
 
 	// For all the IP addresses assigned by IPAM, we will send either a GARP (IPv4) or Unsolicited NA (IPv6).
-	for _, ipc := range ipConfigs {
-		var err error
-		if IsIPv6(ipc.Address.IP) {
-			/* As per RFC 4861, sending unsolicited neighbor advertisements should be considered as a performance
-			* optimization. It does not reliably update caches in all nodes. The Neighbor Unreachability Detection
-			* algorithm is more reliable although it may take slightly longer to update.
-			 */
-			err = SendUnsolicitedNeighborAdvertisement(ipc.Address.IP, linkObj)
-		} else if IsIPv4(ipc.Address.IP) {
-			err = SendGratuitousArp(ipc.Address.IP, linkObj)
-		} else {
-			return fmt.Errorf("the IP %s on interface %q is neither IPv4 or IPv6", ipc.Address.IP.String(), ifName)
-		}
+	//
+	// Note that SendUnsolicitedNeighborAdvertisement() can wait for tentative
+	// IPv6 addresses. It is thus slower. We iterate over the addresses twice
+	// (i_af loop), and first announce IPv4 addresses.
+	for i_af := 0; i_af < 2; i_af++ {
+		for _, ipc := range ipConfigs {
+			var err error
+			if IsIPv6(ipc.Address.IP) {
+				/* As per RFC 4861, sending unsolicited neighbor advertisements should be considered as a performance
+				* optimization. It does not reliably update caches in all nodes. The Neighbor Unreachability Detection
+				* algorithm is more reliable although it may take slightly longer to update.
+				 */
+				if i_af == 0 {
```

Comment:
This looks awkward. Why do we need to do this?
I don't quite understand why we need to loop twice?

Reply:
to apply IPv4 addresses first (during the first iteration).

Reply:
to quote the commit message:
```
Announce fir IPv4 addresses (GARP) before IPv6 NA

To announce IPv6 addresses, SendUnsolicitedNeighborAdvertisement()
possibly needs to wait for tentative IPv6 addresses to become usable
(DAD). This can take more than a second.

Change the loop over all addresses, to iterate them twice, and on the
first run announce IPv4 addresses. IPv4 addresses  don't need to be
affected by this waiting.

Note that utils.AnnounceIPs() is called towards the end of whatever the
command invocation is doing. This reordering, ensures that we
potentially announce IPv4 addresses earlier.
```
Is that unclear?

Reply:
Ah that's the problem. Yes the commit message makes it clear. But in code, as a reader, you don't really know what is going on. It may not be easy to look back in history esp. when the code gets refactored or moved.
This change is a hopeful optimization where there is a chance when doing the GARP for IPv4, it gives enough time for IPv6 DAD to finish.
This optimization comes at a cost of readability of the code. Maybe it would be better to make it more explicit. By having 2 for loops instead of 1 nested for loop.

Reply:
reworked the branch. Better now?

Reply:
Looks good!